// babel-plugin-transform-react-qa-classes/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

exports.default = function (_ref2) {
  var t = _ref2.types;

  return {
    visitor: {
      TaggedTemplateExpression: _styledComponents2.default,
      FunctionDeclaration: function FunctionDeclaration(path, state) {
        if (!path.node.id || !path.node.id.name) return;

        var options = (0, _options2.default)(state);
        var componentName = path.node.id.name;

        functionBodyPushAttributes(t, path, options, componentName);
      },
      ArrowFunctionExpression: function ArrowFunctionExpression(path, state) {
        var options = (0, _options2.default)(state);
        if (!path.parent.id || !path.parent.id.name) return;
        var componentName = path.parent.id.name;

        functionBodyPushAttributes(t, path, options, componentName);
      },
      ClassDeclaration: function ClassDeclaration(path, state) {
        var name = path.get('id');
        var properties = path.get('body').get('body');

        var render = properties.find(function (prop) {
          return prop.isClassMethod() && prop.get('key').isIdentifier({ name: 'render' });
        });

        if (!render || !render.traverse) {
          return;
        }

        var options = (0, _options2.default)(state);

        render.traverse({
          ReturnStatement: function ReturnStatement(returnStatement) {
            var arg = returnStatement.get('argument');
            if (!arg.isJSXElement()) return;

            var openingElement = arg.get('openingElement');

            applyAttribute({
              openingElement: openingElement,
              t: t,
              name: name.node && name.node.name,
              options: options
            });
          }
        });
      }
    }
  };
};

var _options = require('./options');

var _options2 = _interopRequireDefault(_options);

var _styledComponents = require('./styled-components');

var _styledComponents2 = _interopRequireDefault(_styledComponents);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function isReactFragment(openingElement) {
  return openingElement.node.name.name === 'Fragment' || openingElement.node.name.type === 'JSXMemberExpression' && openingElement.node.name.object.name === 'React' && openingElement.node.name.property.name === 'Fragment';
}

function applyAttribute(_ref) {
  var openingElement = _ref.openingElement,
      t = _ref.t,
      name = _ref.name,
      options = _ref.options;

  if (!openingElement || isReactFragment(openingElement)) return;

  var isAttributeAlreadySet = openingElement.node.attributes.find(function (node) {
    if (!node.name) return;
    return node.name.name === options.attribute;
  });

  if (isAttributeAlreadySet) return;

  openingElement.node.attributes.push(t.jSXAttribute(t.jSXIdentifier(options.attribute), t.stringLiteral(options.format(name))));
}

function functionBodyPushAttributes(t, path, options, componentName) {
  var openingElement = null;
  var functionBody = path.get('body').get('body');
  if (functionBody.parent && functionBody.parent.type === 'JSXElement') {
    var jsxElement = functionBody.find(function (c) {
      return c.type === 'JSXElement';
    });
    if (!jsxElement) return;
    openingElement = jsxElement.get('openingElement');
  } else {
    var returnStatement = functionBody.find(function (c) {
      return c.type === 'ReturnStatement';
    });
    if (!returnStatement) return;

    var arg = returnStatement.get('argument');
    if (!arg.isJSXElement()) return;

    openingElement = arg.get('openingElement');
  }

  applyAttribute({ openingElement: openingElement, t: t, name: componentName, options: options });
}
