// eslint-plugin-ui-testing/rules/no-disabled-tests.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RULE_NAME = void 0;
var utils_1 = require("../utils/utils");
exports.RULE_NAME = __filename.slice(__dirname.length + 1, -3);
exports.default = utils_1.createRule({
    name: exports.RULE_NAME,
    meta: {
        docs: {
            category: 'Best Practices',
            description: 'Disallow disabled tests',
            recommended: 'warn',
        },
        messages: {
            noDisabledTests: 'Avoid disabled tests',
        },
        schema: [],
        type: 'suggestion',
    },
    defaultOptions: [],
    create: function (context) {
        return {
            'CallExpression[callee.object.name=/^(describe|it|test)$/] Identifier[name=/^(skip)$/]': function rule(node) {
                context.report({ node: node, messageId: 'noDisabledTests' });
            },
            'CallExpression[callee.object.object.name=/^(describe|it|test)$/] Identifier[name=/^(skip)$/]': function rule(node) {
                context.report({ node: node, messageId: 'noDisabledTests' });
            },
            'CallExpression[callee.name=/^(xdescribe|xit|xtest)$/]': function rule(node) {
                context.report({ node: node, messageId: 'noDisabledTests' });
            },
        };
    },
});
