// eslint-plugin-ui-testing/rules/no-implicit-wait.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RULE_NAME = void 0;
var utils_1 = require("../utils/utils");
exports.RULE_NAME = __filename.slice(__dirname.length + 1, -3);
exports.default = utils_1.createRule({
    name: exports.RULE_NAME,
    meta: {
        docs: {
            category: 'Best Practices',
            description: 'Disallow implicit wait',
            recommended: 'error',
        },
        messages: {
            noImplicitWait: 'Avoid implicit wait, use explicit wait instead',
        },
        schema: [],
        type: 'problem',
    },
    defaultOptions: [],
    create: function (context) {
        return {
            'CallExpression[callee.object.name=/^(browser)$/][callee.property.name=/^(setTimeout)$/] Identifier[name=\'implicit\']': function rule(node) {
                context.report({
                    node: node,
                    messageId: 'noImplicitWait',
                });
            },
        };
    },
});
