// eslint-plugin-ui-testing/rules/no-link-text-selector.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RULE_NAME = void 0;
var utils_1 = require("../utils/utils");
exports.RULE_NAME = __filename.slice(__dirname.length + 1, -3);
exports.default = utils_1.createRule({
    name: exports.RULE_NAME,
    meta: {
        docs: {
            category: 'Best Practices',
            description: 'Disallow link text selector',
            recommended: 'error',
        },
        messages: {
            noLinkTextSelector: 'Avoid link text selector, use css selector instead',
        },
        schema: [],
        type: 'problem',
    },
    defaultOptions: [],
    create: function (context) {
        return {
            'CallExpression[callee.name=/^[$]$/] Literal[value]': function rule(node) {
                var linkTextPattern = new RegExp('^((=)|(\\*=))');
                if (linkTextPattern.test("" + node.value)) {
                    context.report({ node: node, messageId: 'noLinkTextSelector' });
                }
            },
        };
    },
});
