// eslint-plugin-ui-testing/rules/no-xpath-selector.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RULE_NAME = void 0;
var utils_1 = require("../utils/utils");
exports.RULE_NAME = __filename.slice(__dirname.length + 1, -3);
exports.default = utils_1.createRule({
    name: exports.RULE_NAME,
    meta: {
        docs: {
            category: 'Best Practices',
            description: 'Disallow xpath selector',
            recommended: 'warn',
        },
        messages: {
            noXpathSelector: 'Avoid xpath selector, use css selector instead',
        },
        schema: [],
        type: 'suggestion',
    },
    defaultOptions: [],
    create: function (context) {
        return {
            'CallExpression[callee.name=/^[$]$/]': function rule(node) {
                var xpathPattern = new RegExp('^([/]|[(]|(../)|(./)|(\\*/))');
                if (xpathPattern.test(utils_1.getArgumentValue(node))) {
                    context.report({
                        node: node,
                        messageId: 'noXpathSelector',
                    });
                }
            },
        };
    },
});
